zkevm-circuits/src/ecc_circuit.rs

```diff
@@ -463,64 +482,107 @@ impl<F: Field, const XI_0: i64> EccCircuit<F, XI_0> {
-        let p_y_is_zero = ecc_chip.field_chip.is_zero(ctx, &point_p.ec_point.y);
-        let q_x_is_zero = ecc_chip.field_chip.is_zero(ctx, &point_q.ec_point.x);
-        let q_y_is_zero = ecc_chip.field_chip.is_zero(ctx, &point_q.ec_point.y);
-        let r_x_is_zero = ecc_chip.field_chip.is_zero(ctx, &point_r.ec_point.x);
-        let r_y_is_zero = ecc_chip.field_chip.is_zero(ctx, &point_r.ec_point.y);
-        let point_p_is_zero = ecc_chip.field_chip.range().gate().and(
-            ctx,
-            QuantumCell::Existing(p_x_is_zero),
-            QuantumCell::Existing(p_y_is_zero),
-        );
-        let point_q_is_zero = ecc_chip.field_chip.range().gate().and(
-            ctx,
-            QuantumCell::Existing(q_x_is_zero),
-            QuantumCell::Existing(q_y_is_zero),
-        );
-        let point_r_is_zero = ecc_chip.field_chip.range().gate().and(
-            ctx,
-            QuantumCell::Existing(r_x_is_zero),
-            QuantumCell::Existing(r_y_is_zero),
-        );
+        EcAddDecomposed {
+            point_p,
+            point_q,
+            point_r,
+        }
+    }
 
-        // sum1 = if P == (0, 0) then r else r + P
-        let sum1 = ecc_chip.add_unequal(ctx, &rand_point, &point_p.ec_point, true);
-        let sum1 = ecc_chip.select(ctx, &rand_point, &sum1, &point_p_is_zero);
+    /// Phase 2 component for decomposing an EcAdd operation to return each G1 element as cells
+    /// representing its byte form.
+    ///
+    /// It performs proves relations of the decomposed and committed points.
+    #[allow(clippy::too_many_arguments)]
+    fn decompose_ec_add_op_phase_2(
+        &self,
+        ctx: &mut Context<F>,
+        ecc_chip: &EccChip<F, FpConfig<F, Fq>>,
+        ec_decomposed: &[EcAddDecomposed<F>],
+        challenges: &Challenges<Value<F>>,
+    ) {
+        log::trace!("[ECC] ==> EcAdd Assignment START:");
+        log_context_cursor!(ctx);
 
-        // sum2 = if Q == (0, 0) then sum1 else sum1 + Q
-        let sum2 = ecc_chip.add_unequal(ctx, &sum1, &point_q.ec_point, true);
-        let sum2 = ecc_chip.select(ctx, &sum1, &sum2, &point_q_is_zero);
+        let rand_point = {
+            // ideally we want to read a random point via
+            //   `ecc_chip.load_random_point::<G1Affine>(ctx);`
+            // but this function uses internal rng which changes during different phases
+            // this results into an error that `PHASE ERR columnXX not same after phase
+            // Phase(1)`
+            //
+            // so we use an rng from challenge here which ensures the point
+            // remain the same during different phases
+
+            let mut buf = F::default();
+            challenges.keccak_input().map(|x| buf = x);
+            let mut rng = ChaCha20Rng::from_seed(buf.to_repr());
+            let p = G1Affine::random(&mut rng);
+            let G1Affine { x, y } = p;
+            let point = ecc_chip.load_private(ctx, (Value::known(x), Value::known(y)));
```

Comment:
When the mock prover is doing 1st synthesis, the rng is seeded with 0. But for 2nd synthesis, it is then seeded with `keccak_input()`.